Zend/Optimizer/zend_func_info.c

```diff
@@ -834,6 +834,72 @@ static const func_info_t func_infos[] = {
 static HashTable func_info;
 ZEND_API int zend_func_info_rid = -1;
 
+void zend_verify_internal_func_info(zend_function *fn, zval *retval) {
+	if (fn->common.scope) {
+		/* This is a method, not a function. */
+		return;
+	}
+
+	zend_string *name = fn->common.function_name;
+	if (!name) {
+		/* The pass function has no name. */
+		return;
+	}
+
+	zval *zv = zend_hash_find_ex(&func_info, name, 1);
+	if (!zv) {
+		return;
+	}
+
+	func_info_t *info = Z_PTR_P(zv);
+	if (info->info_func) {
+		return;
+	}
+
+	uint32_t type_mask = info->info;
+	if (Z_REFCOUNTED_P(retval)) {
+		if (!(type_mask & MAY_BE_RC1)) {
+			zend_error_noreturn(E_CORE_ERROR, "%s() missing rc1", ZSTR_VAL(name));
+		}
+		if (Z_REFCOUNT_P(retval) > 1 && !(type_mask & MAY_BE_RCN)) {
+			zend_error_noreturn(E_CORE_ERROR, "%s() missing rcn", ZSTR_VAL(name));
+		}
+	}
+
+	uint32_t type = 1u << Z_TYPE_P(retval);
+	if (!(type_mask & type)) {
+		zend_error_noreturn(E_CORE_ERROR, "%s() missing type", ZSTR_VAL(name));
```

Comment:
Bug reports may be more actionable with the value of the type

```suggestion
		zend_error_noreturn(E_CORE_ERROR, "%s() missing type %d", ZSTR_VAL(name), (int) Z_TYPE_P(retval));
```